### PR TITLE
support Graphql Logic endpoint

### PR DESCRIPTION
**Pull request by IanCao**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

File: apm-sniffer/apm-sdk-plugin/graphql-plugin/graphql-12.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/graphql/v12/GraphqlInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        ExecutionStrategyParameters parameters = (ExecutionStrategyParameters) allArguments[1];
+        if (parameters == null || parameters.getParent().getPath() != ExecutionPath.rootPath()) {
+            return ret;
+        }
+        String name = parameters.getField().getSingleField().getName();
+        long latency = System.currentTimeMillis() - (long) objInst.getSkyWalkingDynamicField();
+        String info = buildLogicEndpointTagInfo(name, latency, null);
+        AbstractSpan span = ContextManager.firstSpan();
+        if (span == null || !span.isEntry()) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/graphql-plugin/graphql-12.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/graphql/v12/GraphqlInterceptor.java:55]: How a GraphQL span could not be the first span? Or not an entry span? A little confused.

**Comment by wu-sheng** [comment]
Plugin test and CI fail, please recheck.

**Comment by kezhenxu94** [comment]
`@author` tag and Chinese comments should be removed

**Comment by IanCao** [comment]
> There're `@author` tags and Chinese comments

I've removed them all

File: apm-sniffer/apm-sdk-plugin/graphql-plugin/graphql-12.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/graphql/v12/GraphqlInterceptor.java (hunk 1 above)
Comment (IanCao) [reply apm-sniffer/apm-sdk-plugin/graphql-plugin/graphql-12.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/graphql/v12/GraphqlInterceptor.java]: i've used a localSpan to represent a Grqphql subquery

File: test/plugin/scenarios/graphql-12.x-scenario/pom.xml [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+    <pluginRepositories>
+        <pluginRepository>
+            <id>spring-snapshots</id>
```
Comment (wu-sheng) [inline test/plugin/scenarios/graphql-12.x-scenario/pom.xml:118]: Why need this?

File: test/plugin/scenarios/graphql-8.x-scenario/pom.xml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+    <pluginRepositories>
+        <pluginRepository>
+            <id>spring-snapshots</id>
```
Comment (wu-sheng) [inline test/plugin/scenarios/graphql-8.x-scenario/pom.xml:117]: Same here.

File: test/plugin/scenarios/graphql-9.x-scenario/pom.xml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+    <pluginRepositories>
+        <pluginRepository>
+            <id>spring-snapshots</id>
```
Comment (wu-sheng) [inline test/plugin/scenarios/graphql-9.x-scenario/pom.xml:117]: Same here.

File: test/plugin/scenarios/graphql-9.x-scenario/config/expectedData.yaml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      spanLayer: Unknown
+      startTime: nq 0
+      endTime: nq 0
+      componentId: 92
+      isError: false
+      spanType: Local
+      peer: ''
+      skipAnalysis: false
+      tags:
+      - {key: x-le, value: '{logic-span=true}'}
```
Comment (kezhenxu94) [inline test/plugin/scenarios/graphql-9.x-scenario/config/expectedData.yaml:35]: Such kind of tags are not processable by the backend, according to [the doc](https://github.com/apache/skywalking/blob/310b8857b2266768ccd9e1338da13c86c313ed3f/docs/en/guides/Java-Plugin-Development-Guide.md#extension-logic-endpoint-tag-key-x-le)

> The value of x-le should be JSON format

@IanCao  did you check this plugin with the backend in a whole? FYI @wu-sheng

File: apm-sniffer/apm-sdk-plugin/graphql-plugin/graphql-9.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/graphql/v9/GraphqlInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    private void dealException(Throwable throwable) {
+        AbstractSpan span = ContextManager.activeSpan();
+        span.errorOccurred();
+        span.log(throwable);
+    }
+
+    private String buildLogicEndpointSpan() {
+        Map<String, Object> logicEndpointSpan = new HashMap<>();
+        logicEndpointSpan.put("logic-span", true);
+        return logicEndpointSpan.toString();
```
Comment (kezhenxu94) [inline apm-sniffer/apm-sdk-plugin/graphql-plugin/graphql-9.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/graphql/v9/GraphqlInterceptor.java:76]: The value of `x-le` should be JSON format, this is obviously not the case by simply calling `toString` on a `Map`

File: test/plugin/scenarios/graphql-9.x-scenario/config/expectedData.yaml (hunk 5 above)
Comment (wu-sheng) [reply test/plugin/scenarios/graphql-9.x-scenario/config/expectedData.yaml:35]: Why not? I added this new feature :) It is here

https://github.com/apache/skywalking/blob/310b8857b2266768ccd9e1338da13c86c313ed3f/docs/en/guides/Java-Plugin-Development-Guide.md#extension-logic-endpoint-tag-key-x-le

Comment (kezhenxu94) [reply test/plugin/scenarios/graphql-9.x-scenario/config/expectedData.yaml:35]: I mean `{logic-span=true}` is not JSON format, did you process this kind of format @wu-sheng ? I mean the `=`, it's not JSON format

Comment (kezhenxu94) [reply test/plugin/scenarios/graphql-9.x-scenario/config/expectedData.yaml:35]: @wu-sheng do you mean this, I didn't see any clue that the map format `{logic-span=true}` is also supported

image

Comment (wu-sheng) [reply test/plugin/scenarios/graphql-9.x-scenario/config/expectedData.yaml:35]: Oh, I see, I didn't check that part.

**Pull request merged**